timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,482 @@
+    const int N,
+    const unsigned int *d_atom_idxs,
+    const double *d_coords,
+    const double *d_box,
+    unsigned int *d_perm,
+    cudaStream_t stream) {
+
+    const int tpb = 32;
+    const int B = ceil_divide(N, tpb);
+
+    k_coords_to_kv_gather<<<B, tpb, 0, stream>>>(
+        N, d_atom_idxs, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
+
+    gpuErrchk(cudaPeekAtLastError());
+
+    cub::DeviceRadixSort::SortPairs(
+        d_sort_storage_,
+        d_sort_storage_bytes_,
+        d_sort_keys_in_,
+        d_sort_keys_out_,
+        d_sort_vals_in_,
+        d_perm,
+        N,
+        0,                            // begin bit
+        sizeof(*d_sort_keys_in_) * 8, // end bit
+        stream                        // cudaStream
+    );
+
+    gpuErrchk(cudaPeekAtLastError());
+}
+
+template <typename RealType, bool Interpolated>
+void NonbondedInteractionGroup<RealType, Interpolated>::execute_device(
+    const int N,
+    const int P,
+    const double *d_x,
+    const double *d_p,   // 2 * N * 3
+    const double *d_box, // 3 * 3
+    const double lambda,
+    unsigned long long *d_du_dx,
+    unsigned long long *d_du_dp,
+    unsigned long long *d_du_dl,
+    unsigned long long *d_u,
+    cudaStream_t stream) {
+
+    // (ytz) the nonbonded algorithm proceeds as follows:
+
+    // (done in constructor), construct a hilbert curve mapping each of the 256x256x256 cells into an index.
+    // a. decide if we need to rebuild the neighborlist, if so:
+    //     - look up which cell each particle belongs to, and its linear index along the hilbert curve.
+    //     - use radix pair sort keyed on the hilbert index with values equal to the atomic index
+    //     - resulting sorted values is the permutation array.
+    //     - permute lambda plane/offsets, coords
+    // b. else:
+    //     - permute new coords
+    // c. permute parameters
+    // d. compute the nonbonded interactions using the neighborlist
+    // e. inverse permute the forces, du/dps into the original index.
+    // f. u and du/dl is buffered into a per-particle array, and then reduced.
+    // g. note that du/dl is not an exact per-particle du/dl - it is only used for reduction purposes.
+
+    if (N != N_) {
+        std::cout << N << " " << N_ << std::endl;
+        throw std::runtime_error("NonbondedInteractionGroup::execute_device() N != N_");
+    }
+
+    const int M = Interpolated ? 2 : 1;
+
+    if (P != M * N_ * 3) {
+        std::cout << P << " " << N_ << std::endl;
+        throw std::runtime_error("NonbondedInteractionGroup::execute_device() P != M*N_*3");
+    }
+
+    // identify which tiles contain interpolated parameters
+
+    const int tpb = 32;
+    const int B = ceil_divide(N_, tpb);
+
+    // (ytz) see if we need to rebuild the neighborlist.
+    // (ytz + jfass): note that this logic needs to change if we use NPT later on since a resize in the box
+    // can introduce new interactions.
```

Comment:
Outdated comment (present also in previous)

Reply:
ah, should this comment be removed both here and in `nonbonded_all_pairs.cu`? It wasn't clear to me that it's outdated

Reply:
I think the comment can be removed both places!

Was addressed last year by triggering nblist rebuild whenever box changes at all https://github.com/proteneer/timemachine/blob/5d8f3099aafa1cf7def0585525aaf4de73c23343/timemachine/cpp/src/kernels/k_nonbonded_dense.cuh#L48-L82

Reply:
makes sense! removed the obsolete comment in both places in 4f4bc51